tests/cpp/test_move_pad.cpp

```diff
@@ -0,0 +1,437 @@
+  TensorView* tv2 = makeContigConcreteTensor({4, 5});
+  // tv1_b [bS, iS]. pad axes [1,] is not on broadcast dimension, we should be
+  // able to propagate pad
+  TensorView* tv3 = add(tv0, tv1_b);
+  TensorView* tv4 = cat({tv2, tv3}, /*dim=*/1);
+
+  fusion->addInput(tv0);
+  fusion->addInput(tv1);
+  fusion->addInput(tv2);
+  fusion->addOutput(tv4);
+
+  auto options = at::TensorOptions().dtype(at::kFloat).device(at::kCUDA, 0);
+  at::Tensor t0 = at::randn({4, 10}, options);
+  at::Tensor t1 = at::randn({10}, options);
+  at::Tensor t2 = at::randn({4, 5}, options);
+  std::vector<c10::IValue> aten_inputs = {t0, t1, t2};
+
+  FusionExecutorCache fec(std::move(fusion));
+  auto out_tensors = fec.runFusionWithInputs(aten_inputs);
+
+  FusionKernelRuntime* runtime = fec.getMostRecentKernelRuntime();
+  EXPECT_EQ(runtime->fusionSegments()->groups().size(), 2);
+
+  // ensure that we propagate the pad across binary operation and the first
+  // segment is no-op
+  auto scheduler = runtime->schedulerHeuristics()->heuristicsList().at(0).get();
+  auto heuristic = scheduler->heuristic();
+  NVF_CHECK(
+      heuristic == ScheduleHeuristic::NoOp,
+      "Unexpected heuristic: ",
+      heuristic);
+
+  testValidate(fec.fusion(), out_tensors, aten_inputs, __LINE__, __FILE__);
+}
+
+TEST_F(MovePadTest, BinaryBroadcastOnCatDim) {
+  auto fusion = std::make_unique<Fusion>();
+  FusionGuard fg(fusion.get());
+
+  TensorView* tv0 = makeContigConcreteTensor({4, 10});
+  TensorView* tv1 = makeContigConcreteTensor({10});
+  TensorView* tv1_b = broadcast(tv1, {true, false});
+  TensorView* tv2 = makeContigConcreteTensor({2, 10});
+  // tv1_b [bS, iS]. pad axes [0,] is includes broadcast dimension, pad
+  // propagation stops here
+  TensorView* tv3 = add(tv0, tv1_b);
+  TensorView* tv4 = cat({tv2, tv3}, /*dim=*/0);
+
+  fusion->addInput(tv0);
+  fusion->addInput(tv1);
+  fusion->addInput(tv2);
+  fusion->addOutput(tv4);
+
+  auto options = at::TensorOptions().dtype(at::kFloat).device(at::kCUDA, 0);
+  at::Tensor t0 = at::randn({4, 10}, options);
+  at::Tensor t1 = at::randn({10}, options);
+  at::Tensor t2 = at::randn({2, 10}, options);
+  std::vector<c10::IValue> aten_inputs = {t0, t1, t2};
+
+  FusionExecutorCache fec(std::move(fusion));
+  auto out_tensors = fec.runFusionWithInputs(aten_inputs);
+
+  FusionKernelRuntime* runtime = fec.getMostRecentKernelRuntime();
+  EXPECT_EQ(runtime->fusionSegments()->groups().size(), 2);
+
+  testValidate(fec.fusion(), out_tensors, aten_inputs, __LINE__, __FILE__);
+}
+
+TEST_F(MovePadTest, PadReplayOnMultipleUsesCase0) {
+  auto fusion = std::make_unique<Fusion>();
+  FusionGuard fg(fusion.get());
+
+  TensorView* tv0 = makeContigConcreteTensor({4, 10});
+  TensorView* tv1 = makeContigConcreteTensor({1, 10});
+  // pad on tv5 will propagation back to tv0, since all its uses are traversed
+  //
+  //   tv0 --> tv2 --> tv3 --> tv5
+  //               \        /
+  //                -> tv4 -
```

Comment:
I'm trying to trace what happens in this example. We start with `pad(tv5)` then we convert that to
```
add(pad(tv3), pad(tv4));
```
Then I think the stack-based approach we use will process `pad(tv4)`, replacing it with `sin(pad(tv2))`, then replacing `pad(tv2)` with `relu(pad(tv0))`. At this point we can't continue but there is another input PadOp on the frontier and that is `pad(neg(pad(tv0)))` (since we have already replaced `tv2`). After one more move the two pads bump into one another; as I understand it this means we "stack" the pads. How are we avoiding doubling the pad widths in this case?

Reply:
Sorry the propagation logic above has a subtle thing I should have highlighted.

we are not supposed to have `pad(neg(pad(tv0)))`.
when we have `add(neg(pad(tv2)), sin(pad(tv2)))`. The next step would try to unify the pad operations on tv2. Look at the use of `isSamePadOp` above. and we'll merge the fork there. Afterwards, the graph becomes:

```
tv0 -> tv2 - (merged_pad) -> tv_new (forked with neg & sin)
```

then we'll propagate the merge pad up to tv0.

We avoid repetitive application of pad on a single tensor with fork by checking the uses of pad->input to be 1 before propagating any pad across (this applies to unary/binary/catOp).

Reply:
How is it that we would get `sin(tv4)` before proceeding though? What stops the frontier propagating past `neg(pad(tv2))`?

Reply:
when we walk up the `neg` branch and reached `neg(pad(tv2))`. If the other `sin` branch is not ready yet (i.e. the pad is still down somewhere like `pad(sin(tv2))`.

At line 357, we'll check the `tv2->uses().size()` and if tv2 has more uses than 1, the propagation will abort on `neg` line.
We would only propagate pad after `tv2`, when it walks up the second time from the `sin` branch. At which point, all uses are consumed by the identical `PadOp` and merged as a single use.

Reply:
I realized I should mention this in the code comment. I'll put it in `Note [ PadOp Propagation Rule ]`

Reply:
Also, please do drop more questions on this if it doesn't sound clear. That'll help me put up a better comment. :bow:

Reply:
Ah that's it! Lines 357 and 387 are the key there. Sounds good. Thanks for the explanation!